enhancements/installer/cluster-api-based-installations.md

```diff
@@ -0,0 +1,347 @@
+---
+title: cluster-api-based-installations
+authors:
+  - "@patrickdillon"
+  - "@vincepri"
+  - "@JoelSpeed"
+reviewers:
+  - "@sdodson"
+  - "@2uasimojo"
+  - "@mhrivnak"
+  - "@jnpacker"
+  - "@dtantsur"
+  - "@mdbooth"
+  - "@rvanderp3"
+  - "@jcpowermac"
+  - "@enxebre"
+  - "@wking"
+  - "@dhellmann"
+approvers:
+  - "@zaneb"
+api-approvers:
+  - "None"
+creation-date: 2023-12-12
+last-updated: 2023-12-12
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/CORS-2840
+see-also:
+  - "/enhancements/this-other-neat-thing.md"
+replaces:
+  - "/enhancements/that-less-than-great-idea.md"
+superseded-by:
+  - "/enhancements/our-past-effort.md"
+---
+
+# Enabling Cluster-API-based Installations via openshift-install
+
+## Summary
+
+This enhancement discusses how `openshift-install` can use
+cluster-api (CAPI) providers to provision cluster infrastructure,
+without requiring access to an external management cluster or a local container runtime.
+By running a Kubernetes control plane and CAPI-provider controllers as
+subprocesses on the installer host, `openshift-install` can use CAPI and its
+providers in a similar manner to how Terraform and its providers are currently
+being used.
+
+## Motivation
+
+1. OpenShift Alignment with CAPI: CAPI offers numerous potential benefits;
+such as: day-2 infrastructure management, an API for users to edit cluster
+infrastructure, and upstream collaboration. Installer support for CAPI would
+be foudational for adopting these benefits.
+
+2. Terraform BSL License Change: due to the restrictive license change of
+Terraform, `openshift-install` needs a framework to replace the primary
+tool it used to provision cluster infrastructure. In addition to the benefits
+listed above, CAPI provides solutions for the biggest gaps left by Terraform:
+a common API for specifying infrastructure across cloud providers and robust
+infrastructure error handling.
+
+3. Streamline Installer development: a common pattern for Installer development
+has been to reimplement control-plane features in Terraform that have already been
+delivered upstream. By utilizing CAPI providers for Day-0 provisioning, our development
+practices will remove this duplication and become more efficient.
+
+### User Stories
+
+- As an existing user/client of the installer, I want backwards compatibility so that I can continue to use the installer (e.g. `create cluster`) in the same manner and with existing automation.
+- As a security analyst, I want the installer image to be free of Terraform and related dependencies to decrease surface area for vulnerabilities.
+- As an advanced user or cluster administrator, I want to be able to edit the CAPI infrastructure manifests so that I can customize control-plane infrastructure.
```

Comment:
Maybe it's a good time stop deleting the install config when installation is triggered

Reply:
> Maybe it's a good time stop deleting the install config when installation is triggered

Seems out of scope for this discussion, but check out https://github.com/openshift/installer/pull/7754

Reply:
Did we support this in the current terraform codebase? If not, are we saying that being able to modify infrastructure resources is a blocker for completing this or a future featureset?

Reply:
We didn't support it, we gave users the choice to do the UPI flow. I don't think this is a blocker, but gives an opportunity for a much better custom infrastructure flow

Reply:
I disagree this doesn't belong in user stories and that this is an implementation detail. It describes the most significant functionality we are introducing to the users. 

> Did we support this in the current terraform codebase? 

This was not possible in terraform, and it introduces the potential of greater configurability with the risk that the inputs are not validated through the install config. They would get webhook validation, but a poor user experience in addition to a failed install is certainly possible. It would give users significantly greater control over the infrastructure provisioned by the installer. 

> If not, are we saying that being able to modify infrastructure resources is a blocker for completing this or a future featureset?

It is a proposed feature of this enhancement, but it is not a necessity. We could gate the manifest creation if it is potentially problematic or not ready. @vincepri wdyt?

Reply:
This was discussed in the arch call this week and the enhancement has been updated to indicate that guidelines for editing manifests will be included in docs, as well as the manifests to be included in the cluster as a configmap.